quickjs.c

```diff
@@ -57422,6 +57423,144 @@ static void insert_weakref_record(JSValueConst target,
     *pwr = wr;
 }
 
+/* urlsafe_base64 atob/btoa */
+
+// Base64 encoding table
+static const char base64_table[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";
+
+// btoa: binary to ASCII (base64 encode)
+char* btoa(JSContext *ctx, const char* bin, size_t len) {
+    if (bin == NULL || len == 0) return NULL;
+    // Calculate output length (including padding)
+    size_t out_len = 4 * ((len + 2) / 3) + 1;
+    // Allocate memory for output string (plus null terminator)
+    char* out = (char*)js_mallocz(ctx, out_len);
+
+    if (out == NULL) return NULL;
+
+    // Base64 encoding process
+    size_t i, j;
+
+    for (i = 0, j = 0; i < len; i += 3, j += 4) {
+        uint32_t triple = (bin[i] << 16);
```

Comment:
```suggestion
        uint32_t triple = (uint8_t)bin[i] << 16;
```
Otherwise if `char` is signed and `bin[i] == -42` (for example), you get unexpected output; plus, shifting into the sign bit is ill-defined behavior in C. If you were to shift by 24 bits, you'd have to write it as `(uint32_t)(uint8_t)bin[i] << 16` (or change `bin` to `uint8_t *`)